core/store/src/trie/trie_storage.rs

```diff
@@ -13,25 +13,120 @@ use std::cell::RefCell;
+    CachingShard,
+    CachingChunk,
+}
+
+pub enum RetrievalCost {
+    Free,
+    Full,
+}
 
 impl TrieCache {
+    fn get_cache_position(&mut self, hash: &CryptoHash) -> CachePosition {
+        match self.chunk_cache.get(hash) {
+            Some(value) => CachePosition::ChunkCache(value.clone()),
+            None => match self.shard_cache.get(hash) {
+                Some(value) => CachePosition::ShardCache(value.clone()),
+                None => CachePosition::None,
+            },
+        }
+    }
+
+    pub fn chargeable_get(&mut self, hash: &CryptoHash) -> (Option<Vec<u8>>, RetrievalCost) {
+        match self.get_cache_position(hash) {
+            CachePosition::None => (None, RetrievalCost::Full),
+            CachePosition::ShardCache(value) => {
+                if let CacheState::CachingChunk = &self.cache_state {
+                    let value = self
+                        .shard_cache
+                        .pop(hash)
+                        .expect("If position is ShardCache then value must be presented");
+                    self.chunk_cache.insert(hash.clone(), value);
+                };
+                (Some(value), RetrievalCost::Full)
+            }
+            CachePosition::ChunkCache(value) => (Some(value), RetrievalCost::Free),
+        }
+    }
+
+    fn put(&mut self, hash: CryptoHash, value: Vec<u8>) {
+        // TODO: put TRIE_LIMIT_CACHED_VALUE_SIZE to runtime config
+        if value.len() >= TRIE_LIMIT_CACHED_VALUE_SIZE {
+            return;
+        }
+
+        if let CacheState::CachingChunk = &self.cache_state {
+            self.shard_cache.pop(&hash);
+            self.chunk_cache.insert(hash, value);
+        } else {
+            if self.chunk_cache.contains_key(&hash) {
+                self.chunk_cache.insert(hash, value);
+            } else {
+                self.shard_cache.put(hash, value);
+            }
+        }
+    }
+
+    fn pop(&mut self, hash: &CryptoHash) -> Option<Vec<u8>> {
+        match self.chunk_cache.remove(hash) {
+            Some(value) => Some(value),
+            None => self.shard_cache.pop(hash),
+        }
+    }
+
+    fn drain_chunk_cache(&mut self) {
+        self.chunk_cache.drain().for_each(|(hash, value)| {
+            self.shard_cache.put(hash, value);
+        });
+    }
+}
+
+impl SyncTrieCache {
     pub fn new() -> Self {
-        Self(Arc::new(Mutex::new(LruCache::new(TRIE_MAX_CACHE_SIZE))))
+        Self(Arc::new(Mutex::new(TrieCache {
+            cache_state: CacheState::CachingShard,
+            shard_cache: LruCache::new(TRIE_MAX_CACHE_SIZE),
+            chunk_cache: Default::default(),
+        })))
     }
 
     pub fn clear(&self) {
-        self.0.lock().expect(POISONED_LOCK_ERR).clear()
+        let mut guard = self.0.lock().expect(POISONED_LOCK_ERR);
+        guard.cache_state = CacheState::CachingShard;
+        guard.shard_cache.clear();
+        guard.chunk_cache.clear();
+    }
+
+    pub fn flip_caching_chunk_state(&self) {
```

Comment:
Hm, so, from perspective of idiomatic Rust which cares about ownership and aliasing, this is as bad a signature as it gets. 

The type of this function tells you "'it's safe to call this concurrently from any thread in any moment, because it won't affect concurrent users". The reality is very much the opposite -- the order in which this method and others are called is very much important, there's a specific ordered protocol where you must first flip, then do some operations, than flip back, and it very much relies on exclusive access. 

The fundamental issue here is that our "read" operation isn't really a read, in a sense that it is not idempotent, and has observable side effects (I think this became the case when we added touching `TouchedNodesCounter` which uses the same faux-atomics, while in fact it requires unique access). In this case we have an atypical cache, which is not transparent to the caller, and where we, on the contrary, care very much about internal operation. From Rust's perspective, we really want to have `&mut self` even for `.get` here. 

An extra smell here is that, I think, we only this extra cache to be in effect for the duration of `apply` method: 

https://github.com/near/nearcore/blob/6f9aea9925812ce17b285ccc0876e17c22e762c2/runtime/runtime/src/lib.rs#L1169-L1180

But we are forced to make it global, and force every client of the trie to use implementation with this extra accounting. This also makes stuff more confusing (hence my question of what the heck is chunk and why we care about it here). At the level of abstraction where we implement this, we have chunks and blocks, and care about them. At the level of abstraction where we use it (`Runtime::apply`), we don't have those, we just have a bunch of transactions/recipts.

<hr/>

I don't have a super specific and confident suggestion for the incremental step to make this better. This is a typical hard case where we want to ship "cool feature", but we have "suboptimal architecture" in place, and really need a "big refactor". But doing refactoring and feature-work at the same time is pushing opposing incentives. 

That being said, I wonder if the following would work. Rather than adding the cache to the trie itself, we can *inject* the cache as a parameter to the `get` method. That way, we can create the cache at the beginning of the `apply` method, and have `&mut` access to it. This *cache* would than be able to do the required accounting. In code,

```rust
pub struct Trie {
    pub(crate) storage: Box<dyn TrieStorage>,
    // NB: no node counter here
}

/// A trait to "decorate" access to storage
pub trait RetrieveRawBytes {
    fn retrieve_raw_bytes(
        &mut self,
        key: &CryptoHash,
        // undecorated access
        thunk: impl FnOnce() -> Result<Vec<u8>, StorageError>,
    ) -> Result<Vec<u8>, StorageError> {
    }
}

/// Trivial decorator, which just calls the thunk directly
impl RetrieveRawBytes for () {
    fn retrieve_raw_bytes(
        &mut self,
        key: &CryptoHash,
        thunk: impl FnOnce(&CryptoHash) -> Result<Vec<u8>, StorageError>,
    ) -> Result<Vec<u8>, StorageError> {
        thunk()
    }
}

impl Trie {
    // lookup delegates to lookup_with with trivial retrieve_bytes
    fn lookup(
        &self,
        root: &CryptoHash,
        key: NibbleSlice<'_>,
    ) -> Result<Option<(u32, CryptoHash)>, StorageError> {
        self.lookup_with(root, key, &mut ())
    }

    // The core of the pattern: the caller can customize the behavior here by
    // injecting a custom RetrieveRawBytes. Note also aliasing interplay:
    // `self` is `&`, only `r` is `&mut`.
    fn lookup_with(
        &self,
        root: &CryptoHash,
        mut key: NibbleSlice<'_>,
        r: &mut impl RetrieveRawBytes,
    ) -> Result<Option<(u32, CryptoHash)>, StorageError> {
        let mut hash = *root;

        loop {
            if hash == Trie::empty_root() {
                return Ok(None);
            }
            // Here we call the decorator, passing in the undecorated thunk to it.
            let bytes = r.retrieve_raw_bytes(&hash, || self.storage.retrieve_raw_bytes(&hash))?;
            ...
        }
    }

    // `lookup` is an internal method, we need to do the similar `foo/foo_with` bifurcation of the whole public API.
    pub fn get(&self, root: &CryptoHash, key: &[u8]) -> Result<Option<Vec<u8>>, StorageError> {
        self.get_with(root, key, &mut ())
    }

    pub fn get_with(
        &self,
        root: &CryptoHash,
        key: &[u8],
        r: &mut impl RetrieveRawBytes,
    ) -> Result<Option<Vec<u8>>, StorageError> { ... }
}

pub struct TrieUpdate {
    pub trie: Rc<Trie>,
    // That's our `RetrieveRawBytes`
    trie_node_cache: TrieNodeCache,
    root: CryptoHash,
    committed: RawStateChanges,
    prospective: TrieUpdates,
}

/// Mediator for trie node access, which counts the number of touched nodes and maintains an extra cache. 
/// Note that we no longer need atomics here, as we gained `&mut`
struct TrieNodeCache {
    n_touched_nodes: u64,
    nodes: HashMap<CryptoHash, Vec<u8>>,
}

impl RetrieveRawBytes for TrieNodeCache {
    fn retrieve_raw_bytes(
        &mut self,
        key: &CryptoHash,
        thunk: impl FnOnce() -> Result<Vec<u8>, StorageError>,
    ) -> Result<Vec<u8>, StorageError> {
        // The actual logic of caching access and counting the misses.
        let res = match self.nodes.entry(*key) {
            Entry::Occupied(entry) => entry.get().clone(),
            Entry::Vacant(entry) => {
                let value = thunk()?;
                self.n_touched_nodes += 1;
                entry.insert(value.clone());
                value
            }
        };
        Ok(res)
    }
}

impl TrieUpdate {
   /// Note that this is `&mut self` now, which makes sense. When you do accesses via `TrieUpdate`, 
   /// you want to count cache misses, and to do that correctly you generally want exclusive access
   pub fn get(&mut self, key: &TrieKey) -> Result<Option<Vec<u8>>, StorageError> {
        let key = key.to_vec();
        if let Some(key_value) = self.prospective.get(&key) {
            return Ok(key_value.value.as_ref().map(<Vec<u8>>::clone));
        } else if let Some(changes_with_trie_key) = self.committed.get(&key) {
            if let Some(RawStateChange { data, .. }) = changes_with_trie_key.changes.last() {
                return Ok(data.as_ref().map(<Vec<u8>>::clone));
            }
        }

        // Here we used extended Trie API by passing `self.trie_node_cache` as the decorator for the underlying node access. 
        self.trie.get_with(&self.root, &key, &mut self.trie_node_cache)
    }
}
```